compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerToGlobalLoads.cpp

```diff
@@ -0,0 +1,338 @@
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Utils/Utils.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/MemRef/Utils/MemRefUtils.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/Dialect/Vector/Transforms/VectorTransforms.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/OpDefinition.h"
+
+#define DEBUG_TYPE "iree-llvmgpu-lower-to-global-loads"
+
+#define LDBG(X) LLVM_DEBUG(llvm::dbgs() << X << "\n")
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_GPULOWERTOGLOBALLOADSPASS
+#include "iree/compiler/Codegen/Common/GPU/Passes.h.inc"
+
+namespace {
+//====---------------------------------------------------------------------===//
+// TODO: Move this to a common place.
+//====---------------------------------------------------------------------===//
+
+static constexpr int kNumBitsPerCopy = 32;
+
+// Slice the tensor into numParts parts, and return the i-th part.
+static std::optional<Value> sliceTensor(RewriterBase &rewriter, Value tensor,
+                                        size_t numParts, Value index) {
+  auto tensorType = cast<RankedTensorType>(tensor.getType());
+  auto outermostDimSize = tensorType.getDimSize(0);
+  if (outermostDimSize % numParts != 0) {
+    return std::nullopt;
+  }
+
+  // Create an extract slice
+  SmallVector<int64_t> newShape(tensorType.getShape());
+  newShape[0] = outermostDimSize / numParts;
+
+  auto loc = tensor.getLoc();
+
+  SmallVector<Value> offsets = {rewriter.create<arith::MulIOp>(
+      loc, index, rewriter.create<arith::ConstantIndexOp>(loc, numParts))};
+
+  SmallVector<int64_t> strides =
+      llvm::to_vector(llvm::reverse(tensorType.getShape().drop_back()));
+  strides.push_back(1);
+
+  auto newSliceType =
+      RankedTensorType::get(newShape, tensorType.getElementType());
+  auto noVals = ValueRange{};
+  return rewriter.create<tensor::ExtractSliceOp>(
+      loc, newSliceType, tensor, offsets, noVals, noVals,
+      /*static_offset =*/ArrayRef<int64_t>{INT64_MIN, 0},
+      /*static_shape =*/newShape, /*static_strides =*/strides);
+}
+
+static std::optional<int64_t> getMemRefTypeNumElements(MemRefType memRefType) {
+  auto shape = memRefType.getShape();
+  if (shape.empty()) {
+    return std::nullopt;
+  }
+  int64_t numElements = 1;
+  for (int64_t dim : shape) {
+    if (dim == ShapedType::kDynamic) {
+      return std::nullopt;
+    }
+    numElements *= dim;
+  }
+  return numElements;
+}
+
+static bool distributeLinalgCopyToThreads(RewriterBase &rewriter,
```

Comment:
Add a doc.

Reply:
Also, this might want to be a `LogicalResult`?